flang/unittests/Optimizer/Builder/ComplexHelperTest.cpp

```diff
@@ -0,0 +1,64 @@
+//===- ComplexExprHelperTest.cpp -- ComplexExprHelper unit tests ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Optimizer/Builder/Complex.h"
+#include "gtest/gtest.h"
+#include "flang/Optimizer/Builder/FIRBuilder.h"
+#include "flang/Optimizer/Support/InitFIR.h"
+#include "flang/Optimizer/Support/KindMapping.h"
+
+struct ComplexHelperTest : public testing::Test {
+public:
+  void SetUp() override {
+    fir::KindMapping kindMap(&context,
+        "i10:80,l3:24,a1:8,r54:Double,c20:X86_FP80,r11:PPC_FP128,"
+        "r12:FP128,r13:X86_FP80,r14:Double,r15:Float,r16:Half,r23:BFloat");
+    mlir::OpBuilder builder(&context);
+    auto loc = builder.getUnknownLoc();
+
+    // Set up a Module with a dummy function operation inside.
+    // Set the insertion point in the function entry block.
+    mlir::ModuleOp mod = builder.create<mlir::ModuleOp>(loc);
+    mlir::FuncOp func = mlir::FuncOp::create(
+        loc, "func1", builder.getFunctionType(llvm::None, llvm::None));
+    auto *entryBlock = func.addEntryBlock();
+    mod.push_back(mod);
+    builder.setInsertionPointToStart(entryBlock);
+
+    fir::support::loadDialects(context);
+    firBuilder = std::make_unique<fir::FirOpBuilder>(mod, kindMap);
+  }
+
+  fir::FirOpBuilder &getBuilder() { return *firBuilder; }
+
+  mlir::MLIRContext context;
+  std::unique_ptr<fir::FirOpBuilder> firBuilder;
+};
+
+TEST_F(ComplexHelperTest, mainFunctions) {
+  auto builder = getBuilder();
+  auto loc = builder.getUnknownLoc();
+  auto complexHelper = fir::factory::ComplexExprHelper(builder, loc);
+  auto ctx = builder.getContext();
+  auto realTy = mlir::FloatType::getF32(ctx);
+  mlir::Value rOne = builder.createRealConstant(loc, realTy, 1u);
+  mlir::Value rTwo = builder.createRealConstant(loc, realTy, 2u);
+  fir::ComplexType cType1 = fir::ComplexType::get(ctx, 4);
+  mlir::Value cVal1 = complexHelper.createComplex(cType1, rOne, rTwo);
+  mlir::Value cVal2 = complexHelper.createComplex(4, rOne, rTwo);
+  EXPECT_TRUE(fir::isa_complex(cVal1.getType()));
+  EXPECT_TRUE(fir::isa_complex(cVal2.getType()));
+  mlir::Value real1 = complexHelper.extractComplexPart(cVal1, /*isImagPart=*/ false);
+  mlir::Value imag1 = complexHelper.extractComplexPart(cVal1, /*isImagPart=*/ true);
+  mlir::Value real2 = complexHelper.extractComplexPart(cVal2, /*isImagPart=*/ false);
+  mlir::Value imag2 = complexHelper.extractComplexPart(cVal2, /*isImagPart=*/ true);
+  EXPECT_EQ(realTy, real1.getType());
```

Comment:
Are you going to add code to exercise getComplexPartType and insertComplexPart?

Reply:
Thanks for pointing this out. I have now added tests for these.